### PR TITLE
Compose sample app

### PR DESCRIPTION
**Pull request by tonidero**
### Description
This PR adds a new sample app `MagicWeatherCompose`, very similar to the existing `MagicWeather` but using Jetpack Compose. Many things are copied over from `MagicWeather` including the actual weather change logic and the purchase SDK setup. It does change how we interact with the SDK from there.

To better understand the code, this app has 2 main screens:
- Main
- Paywall

The `Main` screen has a bottom navigation view and has 2 subscreens, `Weather` and `User`.

This can be a better sample for more modern android development. Things remaining:
- [ ] Build app on every PR, to make sure it compiles.
- [ ] Improve UI
- [ ] Use coroutines
- [ ] Add some global state instead of always performing `getCustomerInfo`, so we can use it to listen to the customer info listener.

https://github.com/RevenueCat/purchases-android/assets/808417/7c7c54a9-a331-4e00-8329-0c5e86f2b1eb

### CODE REVIEW

File: examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/utils/PackageExtensions.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import com.revenuecat.purchases.ProductType
+
+val Package.buttonText: String
+    get() {
+        with(product) {
+            return if (type == ProductType.SUBS) {
+                "${price.formatted} for ${period?.value} ${period?.unit?.name?.lowercase()}"
+            } else {
+                "${price.formatted} one time"
+            }
```
[the remark this answers is not in the document]
Comment (aboedo) [reply examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/utils/PackageExtensions.kt:13]: we can always iterate as we feel the need

**Review by aboedo — APPROVED** [review]

File: examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/ui/theme/Theme.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    tertiary = Pink40,
+
+    /* Other default colors to override
+    background = Color(0xFFFFFBFE),
+    surface = Color(0xFFFFFBFE),
+    onPrimary = Color.White,
+    onSecondary = Color.White,
+    onTertiary = Color.White,
+    onBackground = Color(0xFF1C1B1F),
+    onSurface = Color(0xFF1C1B1F),
```
[the remark this answers is not in the document]
Comment (tonidero) [reply examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/ui/theme/Theme.kt:36]: This is part of the generated template app, in case you want to customize it 😅. Didn't dedicate much time to leave it pretty though.

File: examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/ui/theme/Type.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        letterSpacing = 0.sp
+    ),
+    labelSmall = TextStyle(
+        fontFamily = FontFamily.Default,
+        fontWeight = FontWeight.Medium,
+        fontSize = 11.sp,
+        lineHeight = 16.sp,
+        letterSpacing = 0.5.sp
+    )
+     */
```
Comment (vegaro) [inline examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/ui/theme/Type.kt:33]: More comments here

File: examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/paywall/PaywallView.kt [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+@Suppress("LongParameterList")
+@Composable
+fun PaywallView(
+    offering: Offering,
+    modifier: Modifier = Modifier,
+    onPurchaseStarted: ((Package) -> Unit)? = null,
+    onPurchaseCompleted: ((CustomerInfo) -> Unit)? = null,
+    onPurchaseCancelled: (() -> Unit)? = null,
+    onPurchaseErrored: ((PurchasesError) -> Unit)? = null,
+    marketingContent: (@Composable () -> Unit)? = null,
```
Comment (vegaro) [inline examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/paywall/PaywallView.kt:37]: Not a blocker, but do you think there's a way we can remove the number of parameters of this function?

File: examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/user/UserViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    fun initiateLogIn() {
+        _uiState.update { it.copy(shouldStartLoginProcess = true) }
+    }
+
+    fun logIn(newUserId: String) {
+        Purchases.sharedInstance.logInWith(
+            appUserID = newUserId,
+            onError = { error ->
+                _uiState.update { it.copy(displayErrorMessage = error.message) }
```
Comment (vegaro) [inline examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/user/UserViewModel.kt:42]: Should we also set shouldStartLoginProcess to false ?

File: examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/utils/EntitlementInfosExtensions.kt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package com.revenuecat.sample.utils
+
+import com.revenuecat.purchases.EntitlementInfos
+
+fun EntitlementInfos.hasActiveEntitlements(): Boolean {
+    return this.active.isNotEmpty()
+}
```
Comment (vegaro) [inline examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/utils/EntitlementInfosExtensions.kt:7]: We could add another one here that gets the active entitlements by doing `active.keys`

File: examples/MagicWeatherCompose/build.gradle [hunk 7]
```diff
@@ -0,0 +1,6 @@
+// Top-level build file where you can add configuration options common to all sub-projects/modules.
+plugins {
+    id 'com.android.application' version '8.0.2' apply false
+    id 'com.android.library' version '8.0.2' apply false
+    id 'org.jetbrains.kotlin.android' version '1.7.20' apply false
+}
```
Comment (vegaro) [inline examples/MagicWeatherCompose/build.gradle:6]: we can use the gradle library catalogs, but we can do it in another PR

**Review by vegaro — APPROVED** [review]

File: examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/paywall/PaywallView.kt (hunk 4 above)
Comment (tonidero) [reply examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/paywall/PaywallView.kt:37]: Well, thinking about it, we definitely should add a listener of some sort, instead of adding all those callbacks. Might do that in a different PR. That would join 4 parameters into 1. As for the others, I think they need to be there, but we can think of ways to refactor.

File: examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/utils/EntitlementInfosExtensions.kt (hunk 6 above)
Comment (tonidero) [reply examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/utils/EntitlementInfosExtensions.kt:7]: Well, I think that makes sense, but we are currently not displaying the active entitlements anywhere, only whether a user has any active entitlements or not. 

We could add a place where it displays the active entitlements, and checks that it has the correct one instead, as we do in the current magic weather, this just seemed a simpler approach, but I'm ok changing it. Thoughts? @vegaro

File: examples/MagicWeatherCompose/build.gradle (hunk 7 above)
Comment (tonidero) [reply examples/MagicWeatherCompose/build.gradle:6]: Yeah I thought about that. Going to make that change in another PR, so we are using catalogs everywhere

Comment (tonidero) [reply examples/MagicWeatherCompose/build.gradle:6]: Done in #1081

File: examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/utils/EntitlementInfosExtensions.kt (hunk 6 above)
Comment (tonidero) [reply examples/MagicWeatherCompose/app/src/main/java/com/revenuecat/sample/utils/EntitlementInfosExtensions.kt:7]: Talked with @vegaro. Done in https://github.com/RevenueCat/purchases-android/pull/1056/commits/5ab1901049e9a511a722d95d12e2ab62b3800b36
